plugins/outputs/opensearch/opensearch.go

```diff
@@ -0,0 +1,451 @@
+	}
+
+	if strings.Contains(templatePattern, "{{") {
+		templatePattern = templatePattern[0:strings.Index(templatePattern, "{{")]
+	}
+
+	if templatePattern == "" {
+		return fmt.Errorf("template cannot be created for dynamic index names without an index prefix")
+	}
+
+	if (o.OverwriteTemplate) || (!templateExists) || (templatePattern != "") {
+		tp := templatePart{
+			TemplatePattern: templatePattern + "*",
+		}
+
+		t := template.Must(template.New("template").Parse(telegrafTemplate))
+		var tmpl bytes.Buffer
+
+		if err := t.Execute(&tmpl, tp); err != nil {
+			return err
+		}
+
+		indexTempReq := opensearchapi.IndicesPutTemplateRequest{
+			Name: o.TemplateName,
+			Body: strings.NewReader(tmpl.String()),
+		}
+		indexTempResp, errCreateTemplate := indexTempReq.Do(ctx, o.osClient.Transport)
+
+		if errCreateTemplate != nil || indexTempResp.StatusCode != 200 {
+			return fmt.Errorf("OpenSearch failed to create index template %s : %w", o.TemplateName, errCreateTemplate)
+		}
+
+		o.Log.Debugf("Template %s created or updated\n", o.TemplateName)
+	} else {
+		o.Log.Debug("Found existing OpenSearch template. Skipping template management")
+	}
+	return nil
+}
+
+func (o *Opensearch) GetReplacementKeys(indexName string, key string, replacement string) (string, []string) {
+	tagKeys := []string{}
+	startKey := "{{" + key
+	startTag := strings.Index(indexName, startKey)
+
+	for startTag >= 0 {
+		endTag := startTag + strings.Index(indexName[startTag:], "}}")
+
+		if endTag < 0 {
+			startTag = -1
+		} else {
+			tagName := indexName[startTag+len(startKey) : endTag]
+
+			var tagReplacer = strings.NewReplacer(
+				startKey+tagName+"}}", replacement,
+			)
+
+			indexName = tagReplacer.Replace(indexName)
+			tagKeys = append(tagKeys, strings.Trim(strings.TrimSpace(tagName), `"`))
+
+			startTag = strings.Index(indexName, startKey)
+		}
+	}
+
+	return indexName, tagKeys
+}
+
+func (o *Opensearch) GetIndexName(indexName string, eventTime time.Time, tagKeys []string, metricTags map[string]string) string {
+	tagValues := []interface{}{}
+
+	for _, key := range tagKeys {
+		if value, ok := metricTags[key]; ok {
+			tagValues = append(tagValues, value)
+		} else {
+			o.Log.Debugf("Tag '%s' not found, using '%s' on index name instead\n", key, o.DefaultTagValue)
+			tagValues = append(tagValues, o.DefaultTagValue)
+		}
+	}
+
+	indexName = fmt.Sprintf(indexName, tagValues...)
+
+	if strings.Contains(indexName, "{{.Time.Format") {
+		updatedIndexName, dateStrArr := o.GetReplacementKeys(indexName, ".Time.Format", "%DATE%")
+		indexName = strings.Replace(updatedIndexName, "%DATE%", eventTime.UTC().Format(dateStrArr[0]), 1)
+	}
+
+	return indexName
+}
+
+func (o *Opensearch) getPipelineName(pipelineInput string, tagKeys []string, metricTags map[string]string) string {
+	if !strings.Contains(pipelineInput, "%") || len(tagKeys) == 0 {
+		return pipelineInput
+	}
+
+	var tagValues []interface{}
+
```

Comment:
You can speedup the code if you have a good estimate for the length of this list. How about
```suggestion
	tagValues := make([]interface{}, 0, len(tagKeys))
```

Reply:
done